### PR TITLE
Read numpy nans/datetimes

### PR DESCRIPTION
**Pull request by sc1f**
This PR improves the Python API's ability to parse Pandas DataFrames, ensuring compatibility with `numpy.nan` and `numpy.datetime64`. 

- Columns that begin with `numpy.nan` will now have their types inferred as `float`, as `numpy.nan` is a float type. If the column contains a value (i.e. `[nan, nan, 1]`) later on, and the value is not of `float` or `integer` type, then the column will be promoted to a string, all `numpy.nan`s will be converted to `None`, and subsequent writes will be converted to string.
- `numpy.nan` serializes to `None` when exported from Perspective.
- Timestamp generation has been moved to the date parser, and `numpy.datetime64` correctly returns Unix timestamps for use within the C++ engine.

### CODE REVIEW

File: python/perspective/perspective/table/_accessor.py [hunk 1]
```diff
@@ -9,8 +9,10 @@
 from perspective.table.libbinding import t_dtype
 try:
     import pandas
+    import numpy
```
Comment (timkpaine) [inline python/perspective/perspective/table/_accessor.py]: we can make numpy a hard requirement

File: python/perspective/perspective/table/_accessor.py [hunk 2]
```diff
@@ -133,12 +135,23 @@ def marshal(self, cidx, ridx, type):
         column_name = self._names[cidx]
         val = self.get(column_name, ridx)
 
-        # parse string dates/datetimes into objects
-        if isinstance(val, str) and type in (t_dtype.DTYPE_DATE, t_dtype.DTYPE_TIME):
+        # first, check for numpy nans without using numpy.isnan as it tries to cast values
+        if isinstance(val, float) and str(val.real) == "nan":
```
Comment (timkpaine) [inline python/perspective/perspective/table/_accessor.py]: `math.isnan`

File: python/perspective/perspective/table/_date_validator.py [hunk 3]
```diff
@@ -21,12 +26,32 @@ def parse(self, str):
...
+            return None
+
+        if numpy is not None and isinstance(d, numpy.datetime64):
+            if str(d) == "NaT":
+                return None
+            # numpy timestamps return in nanoseconds - reduce to milliseconds
+            return round(d.astype(datetime) / 1000000)
+
+        # python datetime timestamp() returns in seconds - expand to milliseconds
+        return round(d.timestamp() * 1000)
```
Comment (timkpaine) [inline python/perspective/perspective/table/_date_validator.py]: we should make this python2 compliant while we're at it

File: python/perspective/perspective/tests/core/test_aggregates.py [hunk 4]
```diff
@@ -5,7 +5,7 @@
 # This file is part of the Perspective library, distributed under the terms of
 # the Apache License 2.0.  The full license can be found in the LICENSE file.
 #
-from mock import patch
+from unittest.mock import patch
```
Comment (timkpaine) [inline python/perspective/perspective/tests/core/test_aggregates.py]: why?

File: python/perspective/perspective/tests/core/test_layout.py [hunk 5]
```diff
@@ -5,7 +5,7 @@
 # This file is part of the Perspective library, distributed under the terms of
 # the Apache License 2.0.  The full license can be found in the LICENSE file.
 #
-from mock import patch
+from unittest.mock import patch
```
Comment (timkpaine) [inline python/perspective/perspective/tests/core/test_layout.py]: why?

File: python/perspective/perspective/src/fill.cpp [hunk 6]
```diff
@@ -267,7 +268,7 @@ _fill_col_int64(t_data_accessor accessor, t_data_table& tbl, std::shared_ptr<t_c
 
             double fval = item.cast<double>();
             if (isnan(fval)) {
-                WARN("Promoting to string");
+                WARN("Promoting " + name + " to string from int64");
```
Comment (timkpaine) [inline python/perspective/perspective/src/fill.cpp]: use warn with format strings

File: python/perspective/perspective/src/fill.cpp [hunk 7]
```diff
@@ -410,7 +411,7 @@ _fill_col_numeric(t_data_accessor accessor, t_data_table& tbl,
                         type = DTYPE_FLOAT64;
                         col->set_nth(i, fval);
                     } else if (isnan(fval)) {
-                        WARN("Promoting to string");
+                        WARN("Promoting column " + name  + " to string from int32");
```
Comment (timkpaine) [inline python/perspective/perspective/src/fill.cpp]: use warn with format strings

File: python/perspective/perspective/src/fill.cpp [hunk 8]
```diff
@@ -421,9 +422,30 @@ _fill_col_numeric(t_data_accessor accessor, t_data_table& tbl,
                     }
                 } break;
                 case DTYPE_FLOAT32: {
+                    bool is_float = py::isinstance<py::float_>(item);
+                    bool is_numpy_nan = is_float && py::str(item).cast<std::string>() == "nan";
+                    if (!is_float || is_numpy_nan) {
+                        WARN("Promoting column " + name  + " to string from float32");
```
Comment (timkpaine) [inline python/perspective/perspective/src/fill.cpp]: use warn with format strings

File: python/perspective/perspective/src/fill.cpp [hunk 9]
```diff
@@ -421,9 +422,30 @@ _fill_col_numeric(t_data_accessor accessor, t_data_table& tbl,
...
+                        return;
+                    }
                     col->set_nth(i, item.cast<float>());
                 } break;
                 case DTYPE_FLOAT64: {
+                    bool is_float = py::isinstance<py::float_>(item);
+                    bool is_numpy_nan = is_float && py::str(item).cast<std::string>() == "nan";
+
+                    if (!is_float || is_numpy_nan) {
+                        WARN("Promoting column " + name  + " to string from float64");
```
Comment (timkpaine) [inline python/perspective/perspective/src/fill.cpp]: use warn with format strings

File: python/perspective/perspective/table/_date_validator.py [hunk 10]
```diff
@@ -5,8 +5,13 @@
 # This file is part of the Perspective library, distributed under the terms of
 # the Apache License 2.0.  The full license can be found in the LICENSE file.
 #
+from datetime import datetime
 from dateutil.parser import parse
 from perspective.table.libbinding import t_dtype
+try:
+    import numpy
```
Comment (timkpaine) [inline python/perspective/perspective/table/_date_validator.py]: numpy is hard dependency

File: python/perspective/perspective/table/_date_validator.py [hunk 11]
```diff
@@ -21,12 +26,32 @@ def parse(self, str):
...
 
+    def to_timestamp(self, d):
+        '''Return an integer that corresponds to the Unix timestamp, i.e. number of milliseconds since epoch.
+
+        This method converts both datetime.datetime and numpy.datetime64 objects.
+        '''
+        if d is None:
+            return None
+
+        if numpy is not None and isinstance(d, numpy.datetime64):
```
Comment (timkpaine) [inline python/perspective/perspective/table/_date_validator.py]: numpy is hard dependency

File: python/perspective/perspective/src/fill.cpp [hunk 12]
```diff
@@ -421,9 +422,30 @@ _fill_col_numeric(t_data_accessor accessor, t_data_table& tbl,
                     }
                 } break;
                 case DTYPE_FLOAT32: {
+                    bool is_float = py::isinstance<py::float_>(item);
+                    bool is_numpy_nan = is_float && py::str(item).cast<std::string>() == "nan";
```
Comment (timkpaine) [inline python/perspective/perspective/src/fill.cpp]: we can do this without constructing a python string, let me look up the details

**Comment by texodus** [comment]
Thanks for the PR!

**Pull request merged**